api/embedded/pillar.go

```diff
@@ -124,3 +128,169 @@ func (p *PillarApi) GetPillarsHistoryByEpoch(epoch uint64, pageIndex, pageSize u
+		types.ZnnTokenStandard,
+		constants.PillarStakeAmount,
+		data,
+	), nil
+}
+
+func (p PillarApi) RegisterLegacy(name string, producerAddress types.Address, rewardAddress types.Address, publicKey string, signature string, giveBlockRewardPercentage uint8, giveDelegateRewardPercentage uint8) (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(
+		definition.LegacyRegisterMethodName,
+		name,
+		producerAddress,
+		rewardAddress,
+		giveBlockRewardPercentage,
+		giveDelegateRewardPercentage,
+		publicKey,
+		signature,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		constants.PillarStakeAmount,
+		data,
+	), nil
+}
+
+func (p PillarApi) UpdatePillar(name string, producerAddress types.Address, rewardAddress types.Address, giveBlockRewardPercentage uint8, giveDelegateRewardPercentage uint8) (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(
+		definition.UpdatePillarMethodName,
+		name,
+		producerAddress,
+		rewardAddress,
+		giveBlockRewardPercentage,
+		giveDelegateRewardPercentage,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		common.Big0,
+		data,
+	), nil
+}
+
+func (p PillarApi) Revoke(name string) (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(
+		definition.RevokeMethodName,
+		name,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		common.Big0,
+		data,
+	), nil
+}
+
+func (p PillarApi) Delegate(name string) (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(
+		definition.DelegateMethodName,
+		name,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		common.Big0,
+		data,
+	), nil
+}
+
+func (p PillarApi) Undelegate() (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(definition.UndelegateMethodName)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		common.Big0,
+		data,
+	), nil
+}
+
+// Common contract methods
+func (p PillarApi) CollectReward() (*nom.AccountBlock, error) {
+	data, err := definition.ABIPillars.PackMethod(definition.CollectRewardMethodName)
+	if err != nil {
+		return nil, err
+	}
+	return template.CallContract(
+		p.client.ProtocolVersion(),
+		p.client.ChainIdentifier(),
+		types.PillarContract,
+		types.ZnnTokenStandard,
+		common.Big0,
+		data,
+	), nil
```

Comment:
This block seems trivially abstractable into a private function

Reply:
private to which package?
I have mostly tried to follow the dart sdk

i'm not sure how much benefit there is to making functions such as CallPillarContract
but maybe some functions such as CallZeroZnnContract would make sense